dandischema/conf.py

```diff
@@ -0,0 +1,66 @@
+# This file defines the configuration for the DANDI schema
+
+from __future__ import annotations
+
+import re
+from typing import Annotated, Optional
+
+from pydantic import StringConstraints
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+_UNVENDORED_ID_PATTERN = r"[A-Z][-A-Z]*"
+_UNVENDORED_DOI_PREFIX_PATTERN = r"10\.\d{4,}"
+
+
+class Config(BaseSettings):
+    """
+    Configuration for the DANDI schema
+
+    Note
+    ----
+        Since this class is subclass of `pydantic.BaseSettings`, each field of an
+        instance of this class can be populated from an environment variable of the
+        same name prefixed with the prefix defined in `model_config` with the name
+        of the environment variable interpreted **case-insensitively**.
+        For details, see https://docs.pydantic.dev/latest/concepts/pydantic_settings/
+    """
+
+    model_config = SettingsConfigDict(env_prefix="dandi_")
+
+    instance_name: Annotated[
+        str, StringConstraints(pattern=rf"^{_UNVENDORED_ID_PATTERN}$")
+    ] = "DANDI-ADHOC"
+    """Name of the DANDI instance"""
+
+    doi_prefix: Optional[
+        Annotated[
+            str, StringConstraints(pattern=rf"^{_UNVENDORED_DOI_PREFIX_PATTERN}$")
+        ]
+    ] = None
+    """
+    The DOI prefix at DataCite
+    """
+
+    @property
+    def id_pattern(self) -> str:
+        """Regex pattern for the prefix of identifiers"""
+        if "instance_name" in self.model_fields_set:
+            return self.instance_name
+
+        # If the instance name is not set,
+        #   we use a pattern for unvendored DANDI instances
+        return _UNVENDORED_ID_PATTERN
```

Comment:
```suggestion
    @property
    def id_pattern(self) -> str:
        """Regex pattern for the prefix of identifiers"""
        return self.instance_name
```

I think we should simplify `id_pattern` to just the above. Otherwise, vendorizing  the CLI according to the JSON serialization of the `Config` object on the server can be overly complicated or inconsistent with the vendorization on the server.

Reply:
please remind me why we need `id_pattern` at all, or in other words -- how is it used?

It seems that current (might have forgot) course of thinking is that every instance has an `instance_name`, which should pass `_UNVENDORED_ID_PATTERN` check at "config specification time". And then we would just base everything against that `instance_name`, as all further schema regexes which use name in them would just use it, right?

Reply:
Your description is what the proposed change above entails.

The current solution, the one before the proposed change,

```py
    @property
    def id_pattern(self) -> str:
        """Regex pattern for the prefix of identifiers"""
        if "instance_name" in self.model_fields_set:
            return self.instance_name
        # If the instance name is not set,
        #   we use a pattern for unvendored DANDI instances
        return _UNVENDORED_ID_PATTERN
```

sets the property `id_pattern` to `_UNVENDORED_ID_PATTERN` if the field `instance_name` is not set, i.e., defaulted to `"DANDI-ADHOC"`, and only sets the property `id_pattern` to `self.instance_name` when the field `instance_name` is explicitly set.

Because the currently definition of `id_pattern` bases not only on the value of `self.instance_name` but also on how the value of `self.instance_name` comes to be, set by user code or defaulted, it creates complications in achieving the same behavior in both the dandi-archive and dandi-cli.

In the proposed change, one may argue that `id_pattern` is unnecessary since it is exactly the same as `self.instance_name`. However, I prefer to keep it because `id_pattern` is not the same as `instance_name` semantically, and it would be nice to have `Config` class to decide what the ID pattern is without relying on some external code.

Reply:
well, I would argue that if we do not have/use pattern here -- we better not pretend that we do!  then all the regexes where such `id_pattern` is used would make more sense to people since they would explicitly just list the `instance_name`.

Reply:
Per our discussion -- Let's move this out to be logic directly within models.py where we have all other logic which is based on config, and here have only what to be user-facing config variables.